Fix bounds parsing in Scipy optimizers and warn when unsupported

edoaltamura opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
Fixes the parsing of bounds to the Scipy optimizers, which previously raised the error below.
```
TypeError: scipy.optimize._minimize.minimize() got multiple values for keyword argument 'bounds'
```
The present PR fixes this behaviour documented in https://github.com/qiskit-community/qiskit-machine-learning/issues/570 and further discussed in https://github.com/qiskit-community/qiskit-algorithms/issues/57.

### Details and comments
This PR contains two updates. The first one gathers the bounds from `_options` or `_kwargs` if present, then copies them ont `bounds`, finally it deletes the key in the dictionaries, avoiding clashes in the https://github.com/edoaltamura/qiskit-algorithms/blob/77abbcbbeedba111cfde75df0cfea2940c9790dc/qiskit_algorithms/optimizers/scipy_optimizer.py#L165. The changes are below:
https://github.com/edoaltamura/qiskit-algorithms/blob/77abbcbbeedba111cfde75df0cfea2940c9790dc/qiskit_algorithms/optimizers/scipy_optimizer.py#L121-L127

The second update returns a warning when trying to assign a non-None bound value with an optimizer method that doesn't support it. The class `QiskitAlgorithmsOptimizersWarning` is defined in:
https://github.com/edoaltamura/qiskit-algorithms/blob/77abbcbbeedba111cfde75df0cfea2940c9790dc/qiskit_algorithms/exceptions.py#L24

For the bounds parsing, the derived class `QiskitAlgorithmsWarning` is defined and used:
https://github.com/edoaltamura/qiskit-algorithms/blob/77abbcbbeedba111cfde75df0cfea2940c9790dc/qiskit_algorithms/exceptions.py#L36

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/qiskit-community/qiskit-algorithms?pullRequest=155) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/qiskit-community/qiskit-algorithms?pullRequest=155) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/qiskit-community/qiskit-algorithms?pullRequest=155) it.</sub>

woodsp-ibm commented:
This will need a test case that used to fail that does not know so it can be tested around things now work as expected and continue to do so. And if (see discussion below) it raises a UserWarning a test to check that happens as expected.

----
**Discussion**:

As the optimizers were historically they called scipy.optimize.minimize - well they still do for the scipy ones. To have these as objects that could be created and passed, rather than have to deal with functions and partials, that would not have fitted with the design when these were first created, values that needed to end up in the `options` field, ie optimizer specific values beyond the set of "standard" values supplied through the parameters to `minimize` were listed as the option fields by the specific optimizer subclass so they could be taken out of the parameters from the optimizer init and bundled into the options. So options is supposed to be only what can go in `options` for scipy. And that is not bounds. Also bounds was not exposed as something that could be directly set by the user when creating an optimizer - more it came through or was derived from the problem. VQE for instance calls the optimizer with bounds as it gets from the ansatz. Now I can see the value in allowing a user to set the bounds - and perhaps if set overrides anything an algorithm might pass.

So to me bounds ought not to be passed in options as that should strictly just be the scipy minimize options. I think options was exposed here as not all possible options had been exposed as parameters, since initially we made a choice on the ones we expected to be used for the problems that were then being addressed. Having bounds as kwargs in the init seems ok, then does one choose to enforce that as the bounds if an algorithm is setting it via some other mechanism, like VQE does, when it calls minimize. Or let bounds on the minimize override that a user might pass in via kwargs - or have another parameter that you can set to say which overrides which.

Last I am not sure I would want to check if bounds is supported and raise a user error (you don;t check other stuff like jac and it seems like if we did it would be a rabbit hole to get lost in). This scipy base class was designed to be more flexible and allow you to pass the `method` which could be a new optimizer added to scipy. In which case the lookup will default to unsupported even if its supported. For the methods we know about of course its possible, my take would be to not to have any checks and let it fail, if that's what happens, in the place that knows best i.e. scipy.

Dealing with the bounds need some thought/discussion - hence the issue raised i.e. #57. Maybe a simple change could tide things over for now and allow something. @Cryoris You were more involved in this ScipyOptimizer than I, any thoughts?

edoaltamura commented:
Thanks, @woodsp-ibm, I'll work on the two tests you suggested.

edoaltamura commented:
In the fork, `options` can accept bounds, however, I could remove it to stay fully consistent with Scipy. I've also allowed the `bounds` argument in `minimize` to override bounds from `__init__`; a test for this is included. When bounds are used together with a method that doesn't support them, I'm raising a warning and setting `bounds=None` - this is a gentler interface than raising an error, but that's possible too.

Cryoris commented:
I agree with @woodsp-ibm above, we should follow SciPy's interface as closely as possible, which includes the same usage of `options`. However, having `kwargs` in the initializer indeed currently enables passing bounds in two places. It would be good to keep this fix of your PR, though I would argue that the bounds passed in `minimize` should not overwrite what was in `kwargs`.

edoaltamura commented:
I have implemented the suggested changes and adapted the bounds parsing to be as close as possible to the Scipy usage. As in Scipy, bounds can only be parsed in the `minimize()` method. If parsed in `__init__()` now exceptions are raised and a message suggests the intended use.

woodsp-ibm commented on qiskit_algorithms/optimizers/scipy_optimizer.py:
```diff
@@ -76,6 +78,20 @@ def __init__(
         self._max_evals_grouped = max_evals_grouped
         self._kwargs = kwargs
 
+        # Initialise bounds and re-allocate if definition in options or kwargs
```
```suggestion
        # Initialize bounds and re-allocate if definition in options or kwargs
```

woodsp-ibm commented on requirements.txt:
```diff
@@ -1,3 +1,4 @@
 qiskit>=0.44
 scipy>=1.4
 numpy>=1.17
+scikit-quant
```
The scikit-quant package is intended as an optional install https://qiskit-community.github.io/qiskit-algorithms/getting_started.html#optional-installs - it came under a varied set of licenses. This will install it for everyone which is not wanted - so this addition needs to go. The package has not been updated in ages and there is an issue open around its removal - tests with that are even skipped for some of them based on numpy as it has never been updated and CI runs on newer numpy.
* #84

edoaltamura replied:
Thanks for explaining, I've now removed it from `requirements.txt`. I had erroneously added it after seeing the skipped tests.

edoaltamura commented:
@woodsp-ibm are there any further changes you would suggest for this PR?

woodsp-ibm commented on qiskit_algorithms/optimizers/scipy_optimizer.py:
```diff
@@ -116,9 +132,27 @@ def minimize(
+            self._bounds = bounds
+
+        # Remove ignored bounds to suppress the warning of scipy.optimize.minimize
+        if self.is_bounds_ignored and self._bounds is not None:
+            warnings.warn(
+                f"Optimizer method {self._method} does not support bounds. Bounds ignored.",
+                QiskitAlgorithmsOptimizersWarning,
+            )
+            self._bounds = None
+        elif not self.is_bounds_ignored and self._bounds is None:
+            warnings.warn(
+                f"Optimizer method {self._method} may require defining bounds. "
+                f"Check the Scipy documentation for more info: "
+                f"https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html",
+                QiskitAlgorithmsOptimizersWarning,
```
This is logic really only applies to methods (scipy optimizers) that are known about here. Given method is not one we know about then the first test could fail when it does in fact support bounds - here it just defaults to is_bounds_ignored if its not one in the set we know about that supports bounds. 

In the second check, since we know what these optimizers are, thats the only way to is_bounds_ignored False  (at least the present set) do any of them mandate bounds. None is interpreted as no bounds which can also be specified explicitly as (None, None) tuples.

edoaltamura replied:
For the first part, would you suggest warning/raising errors if the method is not among the known ones, before going through the bound-supported logic?

edoaltamura replied:
@woodsp-ibm after thinking about it, I see your point. The logic based on `is_bounds_ignored` is a bit loose in checking against bounds, and perhaps should be delegated to a separate discussion if need be.
To keep this PR in scope with resolving https://github.com/qiskit-community/qiskit-machine-learning/issues/570, I will remove these checks and related tests.

woodsp-ibm commented on test/optimizers/test_optimizers.py:
```diff
@@ -183,6 +184,85 @@ def callback(x):
+            optimizer = SciPyOptimizer("SLSQP")
+
+            # Ensure that _bounds attribute is initially present and =None
+            self.assertTrue(hasattr(optimizer, "_bounds"))
+            self.assertTrue(optimizer._bounds is None)
```
Unit tests are to test public API not internal implementation details which we are arguably free to change as long as it does not affect public API. For me the test would be to place bounds in options or not and check the resultant behavior the end user sees, an error/warning or not happens etc. What happens to internal variables/state we arguably shouldn't care as long as the overall behavior, via the public API, is as expected.

edoaltamura replied:
I agree with this; I removed these two checks and kept the ones that test for placing the bounds.

woodsp-ibm reviewed (commented):
I am not sure of the intent here. I though initially it was to add some way for an end user to pass bounds into the objects for when used with algorithms that take an optimizer since its the algorithm in this case that will be calling minimize. For these algs, like VQE they currently use the ansatz to inform them of any parameter ranges, ie, bounds and pass such to the minimize.

This just seems to add checks into the path - something we had before left to scipy optimizers to deal with accordingly - and normally something we leave to the subsystem that knows best rather than trying to know about it and things changing. The ML issue was the person had tried to add it and then sees failures, its not supported in options and if you ad it in kwargs it ends up being there twice. This attempts to clarify the failure, but without some way of a user passing bounds its less than clear to me. Pass some other field in options that you misspell and it hence unsupported - scipy will be dealing with that however it does it.

If we are to allow bounds to be passed in, the Julien suggested they should not override what an algorithm sets. Perhaps it should be an additional boolean to allow just that though otherwise for algos that set the bounds you will never have a way to override the bounds the algo is using (well actually there is today you would overwrite parameter_bounds in the ansatz, at least for VQE, as that is where it gets its bounds info from - but of course its obscure and known by few.)

edoaltamura commented:
I have applied the suggested changes and the PR is now ready. What's your feedback, and what are the next steps?

woodsp-ibm reviewed (commented):
Options was always meant to mean options as per the scipy optimizer in question. I guess if people are trying to get bounds in through there - maybe the options docstring could be clearer. Do we want to detect/add a warning since its a common error - is it that common? If you did that directly to scipy and added it in the options that is how it would fail. I guess you are unlikely to do that since its mnimize exposes the bounds same as we do here. It just that when constructing one of these that no bound can be specified at that point and bounds comes down to be supplied by whatever algo is calling minimize however it gets bounds.

I will not that the base class has a set_options so even with the check here, done in constructor, if it were set that way it would fail since options is only checked when first passed there and not if updated.

woodsp-ibm commented:
Whatever this nets out as, since its a change that affects behavior from a end user perspective - well at least in terms of the error they get is more meaningful, hopefully, this should have a release note too I think.

edoaltamura commented on qiskit_algorithms/optimizers/scipy_optimizer.py:
```diff
@@ -76,6 +76,20 @@ def __init__(
         self._max_evals_grouped = max_evals_grouped
         self._kwargs = kwargs
 
+        # Initialize bounds and re-allocate if definition in options or kwargs
+        self._bounds: list[tuple[float, float]] | None = None
```
Since I removed the logic that checks for compatibility (turned out not to be robust enough to justify including it) this is also not necessary, and I will remove it in the upcoming commit.

edoaltamura commented on qiskit_algorithms/optimizers/scipy_optimizer.py:
```diff
@@ -116,9 +130,16 @@ def minimize(
+
+        # Allocate bounds attribute
+        if bounds is not None:
+            self._bounds = bounds
```
Same as the comment in __init__, since we don't run checks at the constructor level, we don't need to store it anymore. So I agree with you and removed this part too in the next commit.

edoaltamura commented:
I agree that `self._bounds` is no longer necessary in this implementation. I've removed it in both instances and added release notes as suggested.
Once the CI has run, this PR is ready for merge. I look forward to your advice on the next steps.
